client/findproviders.go

```diff
@@ -1,135 +1,129 @@
 package client
 
 import (
-	"bytes"
 	"context"
-	"encoding/json"
-	"errors"
-	"io"
-	"net/http"
-	"net/url"
 
 	"github.com/ipfs/go-cid"
-	"github.com/ipfs/go-delegated-routing/parser"
+	proto "github.com/ipfs/go-delegated-routing/gen/proto"
 	logging "github.com/ipfs/go-log"
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/multiformats/go-multiaddr"
 )
 
-var logger = logging.Logger("delegated/client")
+var logger = logging.Logger("service/client/delegatedrouting")
 
-func (c *client) FindProviders(ctx context.Context, cid cid.Cid) ([]peer.AddrInfo, error) {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	ch, err := c.FindProvidersAsync(ctx, cid)
+type DelegatedRoutingClient interface {
+	FindProviders(ctx context.Context, key cid.Cid) ([]peer.AddrInfo, error)
+	FindProvidersAsync(ctx context.Context, key cid.Cid) (<-chan FindProvidersAsyncResult, error)
+	GetIPNS(ctx context.Context, id []byte) ([][]byte, error)
```

Comment:
This should be validated if it's going to substitute into the `ValueStore` interface and run alongside the DHT and PubSub in go-ipfs.

See https://github.com/ipfs/go-namesys/blob/bf4b3cffd9bdc0fb668bb6879f17070128145798/routing.go#L82 which indicates that the underlying system is responsible for both validating and only returning increasingly recent records.

DHT Get Validation: https://github.com/libp2p/go-libp2p-kad-dht/blob/0b7ac010657443bc0675b3bd61133fe04d61d25b/routing.go#L320

DHT Get Ordering (so that in the async case only increasingly good records are returned): https://github.com/libp2p/go-libp2p-kad-dht/blob/0b7ac010657443bc0675b3bd61133fe04d61d25b/routing.go#L228.

---

If we weren't going to validate here then we'd need really big warning lights like calling the function `GetIPNSUnvalidated` or something to prevent mistakes, however it's easy enough to add the validators from go-ipns so why not 🤷.

Reply:
@aschmahmann Should the ipns Validator be passed as an option when a delegated routing client is created?

Reply:
It could, although realistically nobody passes in `KeyBooks` anymore (https://github.com/ipfs/go-ipns/blob/33ef6df0c42c6f1c736025dfa2ceecbd064e9dff/record.go#L31) so you could just construct an `ipns.Validator{}` here without any options.